lib/Dialect/AIEX/Transforms/AIETransformBfpTypes.cpp

```diff
@@ -0,0 +1,236 @@
+                      -> std::optional<AIEObjectFifoSubviewType> {
+      auto newElementType = convertType(objectFifoSubviewType.getElementType());
+      if (!newElementType) {
+        llvm::errs()
+            << "Failed to convert ObjectFifoSubviewType element type\n";
+        return nullptr;
+      }
+
+      if (auto newMemRef = dyn_cast<MemRefType>(newElementType))
+        return AIEObjectFifoSubviewType::get(objectFifoSubviewType.getContext(),
+                                             newMemRef);
+
+      llvm::errs()
+          << "ObjectFifoSubviewType element type is not a MemRefType\n";
+      return nullptr;
+    });
+
+    // Add a conversion for FunctionType
+    addConversion([&](FunctionType funcType) -> std::optional<FunctionType> {
+      llvm::SmallVector<Type> newInputTypes;
+      auto check = convertTypes(funcType.getInputs(), newInputTypes);
+      if (check.failed()) {
+        llvm::errs() << "Failed to convert function input types\n";
+        return nullptr;
+      }
+
+      llvm::SmallVector<Type> newOutputTypes;
+      check = convertTypes(funcType.getResults(), newOutputTypes);
+      if (check.failed()) {
+        llvm::errs() << "Failed to convert function output types\n";
+        return nullptr;
+      }
+
+      return FunctionType::get(funcType.getContext(), newInputTypes,
+                               newOutputTypes);
+    });
+
+    // Add conversions for other types as needed (llvm arrays?)
+  }
+};
+
+class BfpToIntegerConversionPattern : public ConversionPattern {
+public:
+  BfpToIntegerConversionPattern(TypeConverter &typeConverter,
+                                MLIRContext *context, bool &conversionFailed)
+      : ConversionPattern(typeConverter, MatchAnyOpTypeTag(), 1, context),
+        conversionFailed(conversionFailed) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    // The objective is to replace all bfp operations by an integer of the
+    // appropriate width. This pass currently does not have any other
+    // functionality.
+
+    // Operation results
+    for (auto result : op->getResults()) {
+      auto conversion = typeConverter->convertType(result.getType());
+      if (!conversion) {
+        conversionFailed = true;
+        return op->emitError()
+               << "Failed to convert result type: " << result.getType();
+      }
+      result.setType(conversion);
+    }
+
+    // Operation operands
+    for (auto operand : op->getOperands()) {
+      auto conversion = typeConverter->convertType(operand.getType());
+      if (!conversion) {
+        conversionFailed = true;
+        return op->emitError()
+               << "Failed to convert operand type: " << operand.getType();
+      }
+      operand.setType(conversion);
+    }
+
+    // Operation attributes
+    // Note: For some reason, the attribute list looks like it is immutable and
```

Comment:
Attributes represent constant data, are unique, and are cached, so they are immutable.